docs/paper/dev/api/inventory-holder.md

```diff
@@ -0,0 +1,159 @@
+```java
+public class MyInventory implements InventoryHolder {
+
+	private final Inventory inventory;
+
+	public MyInventory(MyPlugin plugin) {
+		// Create an Inventory with 9 slots, `this` here is our InventoryHolder.
+		this.inventory = plugin.getServer().createInventory(this, 9);
+	}
+
+	@Override
+	public Inventory getInventory() {
+		return this.inventory;
+	}
+
+}
+```
+
+## Opening the inventory
+
+To open the inventory, first we have to instantiate our `MyInventory` class and then open the inventory for the player.
+You can do that wherever you need.
+
+:::note
+
+We pass an instance of our plugin's main class as it's required by the constructor. If you've used the static method and removed the constructor
+argument you don't have to pass it here.
+
+:::
+
+```java
+Player player; // Assume we have a Player instance.
+               // This can be a command, another event or anywhere else you have a Player.
+
+MyInventory myInventory = new MyInventory(myPlugin);
+player.openInventory(myInventory.getInventory());
+```
+
+## Listening to an event
+
+Once we have the inventory open, we can listen to any inventory events we like and check if the `Inventory#getHolder()`
+returns an instance of our `MyInventory`.
+
+```java
+@EventHandler
+public void onInventoryClick(InventoryClickEvent event) {
+	Inventory inventory = event.getInventory();
+	// Check if the holder is our MyInventory,
+	// if yes, use instanceof pattern matching to store it in a variable immediately.
+	if(!(inventory.getHolder() instanceof MyInventory myInventory)) {
+		// It's not our inventory, ignore it.
+		return;
+	}
+
+	// Do what we need in the event.
+}
+```
+
+## Storing data on the InventoryHolder
+
+You can store extra data for your inventories on the InventoryHolder by adding fields and methods to your class.
+
+Let's make an inventory that counts the amount of times we clicked a stone inside it.
+First let's modify our `MyInventory` class a little:
+
+```java
+public class MyInventory implements InventoryHolder {
+
+	private final Inventory inventory;
+
+	private int clicks = 0; // Store the amount of clicks.
+
+	public MyInventory(MyPlugin plugin) {
+		this.inventory = plugin.getServer().createInventory(this, 9);
+
+		// Set the stone that we're going to be clicking.
+		this.inventory.setItem(0, new ItemStack(Material.STONE));
+	}
+
+	// A method we will call in the listener whenever the player clicks the stone.
+	public void addClick() {
+		this.clicks++;
+		this.updateCounter();
+	}
+
+	// A method that will update the counter item.
+	private void updateCounter() {
+		this.inventory.setItem(8, new ItemStack(Material.BEDROCK, this.clicks));
+	}
+
+	@Override
+	public Inventory getInventory() {
+		return this.inventory;
+	}
+
+}
+```
+
+Now, we can modify our listener to check if the player clicked the stone, and if so, add a click.
+
+```java
+@EventHandler
+public void onInventoryClick(InventoryClickEvent event) {
+	Inventory inventory = event.getInventory();
```

Comment:
There's a little side effect using getInventory instead of getClickedInventory that means when the player will click to a stone in its own inventory (not in your custom inv), the counter will still increase.

Reply:
This was of course not the intended purpose, but perhaps it would be better to either leave that to figure out for the reader themselves or further explain that in some other series about inventories similar to how `Event API` currently is

Reply:
Hmm, i personally think that clicked inventory is more explicitly easy to understand that inventory (that currently always get the top inventory). But just want to make sure you were aware of that, i also think that a doc shouldn't lead to unexpected behavior that differ to what the doc said previously (in this case "the amount of times we clicked a stone **inside it**"). But i would be curious to see what others thinks

Reply:
Since there aren't any other opinions and I agree that we should keep to what was said committing with the `getClickedInventory()` now

Reply:
There currently an NPE when you click outside of the window i think a (inventory == null || !(inventory.getHolder() instanceof MyInventory myInventory) should be added

Reply:
I highly doubt instanceof throws NPE on null

Reply:
No it doesn't but the inventory var could be null in this case you call the getHolder on a null pointer

Reply:
Ah, got it